src/meson.build

```diff
@@ -0,0 +1,100 @@
+ffi_c_sources = [
+  'prep_cif.c',
+  'types.c',
+  'raw_api.c',
+  'java_raw_api.c',
+  'closures.c',
+  'tramp.c',
+]
+
+ffi_asm_sources = []
+
+if get_option('ffi-debug')
+  ffi_c_sources += ['debug.c']
+endif
+
+# Translate source files to source paths
+foreach f : c_sources
+  ffi_c_sources += targetdir / f
+endforeach
+foreach f : asm_sources
+  ffi_asm_sources += targetdir / f
+endforeach
+
+ffiinc = include_directories('..', '../include', targetdir)
+
+if is_msvc_like
+  # MSVC, cannot compile assembly files directly. They need to be preprocessed
+  # with the C compiler first, then compiled with MASM.
+  if meson.version().version_compare('<0.64.0')
+    error('Meson >= 0.64 is required with MSVC')
+  endif
+  ffi_asm_sources = cc.preprocess(ffi_asm_sources,
+    output: '@BASENAME@.masm',
+    include_directories: ffiinc,
+  )
+  add_languages('masm', native: false)
+endif
+
+ffi_link_args = []
+ffi_map_file = []
+
+if get_option('ffi-build-versioned')
+  libffi_map_in = meson.project_source_root() / 'libffi.map.in'
+  arg_prefix = '-Wl,--version-script='
+  if cc.has_multi_link_arguments(['-shared', arg_prefix + libffi_map_in])
+    if meson.version().version_compare('>=0.64.0')
+      ffi_map_file = cc.preprocess(libffi_map_in,
+        output: '@BASENAME@',
+        include_directories: ffiinc,
+        compile_args: '-DGENERATE_LIBFFI_MAP',
+      )
+      ffi_map_abs = ffi_map_file[0].full_path()
+    else
+      ffi_map_file = configure_file(
+        output: 'libffi.map',
+        command: [
+          meson.get_compiler('c'),
+          get_option('c_args'),
+          '-I' + join_paths(meson.current_build_dir(), '..'),
+          '-I' + join_paths(meson.current_build_dir(), '..', 'include'),
```

Comment:
While using this branch to unblock some cross-compiling I had trouble doing without it (thanks for that!), I had a trouble in here.

This seems to be missing adding the path to the `targetdir`, which is also added in line 49 above as part of `ffiinc`. As a result, the setup process was failing to find the `ffitarget.h` file in `mips/` (that was in my case).

In my case I sorted it out by just using a more recent version of `meson` so that line 49 was used instead, but probably all that's needed is to add one more include line here (I'm totally new to meson).